numba/targets/arraymath.py

```diff
@@ -668,6 +668,83 @@ def _trivial_isnan(x):
         return _trivial_isnan
 
 
+@overload(np.iscomplex)
+def np_iscomplex(x):
+
+    if not type_can_asarray(x):
+        raise TypingError("First argument must be array-like")
+
+    dt = determine_dtype(x)
+    if np.issubdtype(dt, np.complexfloating):
+        def impl(x):
+            ax = np.asarray(x) # NumPy uses asanyarray here!
+            return ax.imag != 0
+    else:
+        def impl(x):
+            ax = np.asarray(x)
+            zeros = np.zeros(ax.shape, dtype=types.boolean)
+            return zeros[()] # hack when x is a scalar value
+    return impl
+
+
+@overload(np.isreal)
+def np_isreal(x):
+    def impl(x):
+        ax = np.iscomplex(x)
+        return np.invert(ax)
+    return impl
+
+
+@overload(np.iscomplexobj)
+def iscomplexobj(x):
+    # Implementation based on NumPy
+    # https://github.com/numpy/numpy/blob/d9b1e32cb8ef90d6b4a47853241db2a28146a57d/numpy/lib/type_check.py#L282-L320
+    dt = determine_dtype(x)
+    if np.issubdtype(dt, np.complexfloating):
+        @register_jitable
+        def impl(x):
+            return True
+    else:
+        @register_jitable
+        def impl(x):
+            return False
+    return impl
+
+
+@overload(np.isrealobj)
+def isrealobj(x):
+    # Return True if x is a not complex type or an array of complex numbers.
+    # Implementation based on NumPy
+    # https://github.com/numpy/numpy/blob/ccfbcc1cd9a4035a467f2e982a565ab27de25b6b/numpy/lib/type_check.py#L290-L322
+    def impl(x):
+        return not np.iscomplexobj(x)
+    return impl
+
+
+@overload(np.isneginf)
+def isneginf(x, out=None):
+
+    if not type_can_asarray(x):
+        raise TypingError("First argument must be array-like")
+
+    if numpy_version <= (1, 12):
+        wrapper = register_jitable(lambda x: np.asarray(x))
+    else:
+        wrapper = register_jitable(lambda x: x[()])
+
+    if is_nonelike(out):
```

Comment:
The branch pruning should make it so this pattern is not needed. Does this work ok (the checks etc are elided)?:
```python
@overload(np.isneginf)
def isneginf(x, out=None):
    def impl(x, out=None):
        x = np.asarray(x)
        if out is None: # this can be determined at compile time
            out = np.zeros(x.shape, dtype=types.boolean)
        np.logical_and(np.isinf(x), np.signbit(x), out)
        return wrapper(out)
    return impl
```

Reply:
Unfortunately, no! Numba breaks because it thinks `out` can be either `array(bool, 0d, C)` or `None`.

A piece of the error report:
```
In definition 0:
    TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Invalid use of Function(<ufunc 'logical_and'>) with argument(s) of type(s): (bool, bool, OptionalType(array(bool, 0d, C)) i.e. the type 'array(bool, 0d, C) or None')
```

Reply:
hmmm, I think this is lack of SSA biting. As `out` ends up being multiply defined a prune cannot take place as it doesn't know if the value is constant. Try this as a workaround?
```python
@overload(np.isneginf)
def isneginf(x, out=None):

    if not type_can_asarray(x):
        raise TypingError("First argument must be array-like")

    if numpy_version <= (1, 12):
        wrapper = register_jitable(lambda x: np.asarray(x))
    else:
        wrapper = register_jitable(lambda x: x[()])

    def impl(x, out=None):
        x = np.asarray(x)
        if out is None:
            newout = np.empty(x.shape, dtype=types.boolean)
        else:
            newout = out
        np.logical_and(np.isinf(x), np.signbit(x), newout)
        return wrapper(newout)
    return impl
```

Reply:
Nope, got the same thing

```
Invalid use of Function(<function isneginf at 0x10a05e400>) with argument(s) of type(s): (float64, array(bool, 1d, C))
 * parameterized
In definition 0:
    TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Cannot unify array(bool, 0d, C) and array(bool, 1d, C) for 'newout', defined at /Users/guilhermeleobas/Documents/GitHub/numba/numba/targets/arraymath.py (738)
...
```

Reply:
I cannot reproduce what you are seeing, this:
```python
from numba import njit
import numpy as np

from numba import types
from numba.extending import overload, register_jitable
from numba.numpy_support import type_can_asarray
from numba.numpy_support import version as numpy_version

@overload(np.isneginf)
def isneginf(x, out=None):

    if not type_can_asarray(x):
        raise TypingError("First argument must be array-like")

    if numpy_version <= (1, 12):
        wrapper = register_jitable(lambda x: np.asarray(x))
    else:
        wrapper = register_jitable(lambda x: x[()])

    def impl(x, out=None):
        x = np.asarray(x)
        if out is None:
            newout = np.zeros(x.shape, dtype=types.boolean)
        else:
            newout = out
        np.logical_and(np.isinf(x), np.signbit(x), newout)
        return wrapper(newout)
    return impl

@njit
def foo(x, out):
    return np.isneginf(x, out), np.isneginf(x, out=out)

z = np.inf
a = np.array([z, 1])

out = np.empty_like(z).astype(np.bool_)
print(foo(z, out))
print(foo(z, out=out))
print(foo(z, out=None))


out = np.empty_like(a).astype(np.bool_)
print(foo(a, out))
print(foo(a, out=out))
print(foo(a, out=None))
```
works fine against current master:
```
$ git log|head -1
commit ad3303c25133d57127abf8be1cd72d6048f45813
```
is there something I'm missing?

Reply:
Try with this input:
`print(foo(np.NINF, np.zeros(1, dtype=np.bool)))`

Reply:
I think that's a NumPy bug? https://docs.scipy.org/doc/numpy-1.17.0/reference/generated/numpy.isneginf.html
> out : ndarray
    A boolean array with the same dimensions as the input.

Reply:
Yes, probably. I will open an issue on the NumPy issue-tracker and remove those tests that do not meet the specification.

Reply:
Thanks, following this up with NumPy is much appreciated.

Reply:
@stuartarchibald, apparently the input above is a valid input. I've changed the code a bit to prevent an unnecessary allocation when `x` is a scalar value. If CI passes, can you please review it?

Reply:
Thanks, it was definitely worth raising this with NumPy. Seems llike NumPy needs a doc fix and we need to make the above work, which is a shame because it means adding more code. Will take a look at the new patches now.

Reply:
Was this the patch https://github.com/numba/numba/pull/4610/commits/c423a0f3b40efdad9fcb49fbdf77756a705f399c ? doesn't seem related?

Reply:
No, it was this one https://github.com/numba/numba/pull/4610/commits/f172e50085f5f0cadacc72a420dc6ef27aad475d